cachi2/core/package_managers/gomod.py

```diff
@@ -0,0 +1,1030 @@
+import functools
+import fnmatch
+import logging
+import os
+import re
+import shutil
+import tempfile
+from datetime import datetime
+from pathlib import Path, PureWindowsPath
+from typing import Dict, Iterable, List, Optional, Tuple
+
+import backoff
+import git
+import semver
+
+from cachi2.core.config import Config, get_worker_config
+from cachi2.core.errors import (
+    CachitoCalledProcessError,
+    FileAccessError,
+    InvalidFileFormat,
+    GoModError,
+    RepositoryAccessError,
+    UnsupportedFeature,
+    ValidationError,
+)
+from cachi2.core.paths import RequestBundleDir
+from cachi2.core.packages_data import PackagesData
+from cachi2.core.utils import run_cmd, load_json_stream
+
+
+logging.basicConfig(format='%(asctime)s %(levelname)s %(message)s', level=logging.DEBUG)
+log = logging.getLogger(__name__)
+
+
+_run_gomod_cmd = functools.partial(run_cmd, exc_msg="Processing gomod dependencies failed")
+_MODULE_VERSION_RE = re.compile(r"/v\d+$")
+
+
+def fetch_gomod_source(request, dep_replacements=None, package_configs=None):
+    """
+    Resolve and fetch gomod dependencies for a given request.
+
+    :param list dep_replacements: dependency replacements with the keys "name" and "version"; only
+        supported with a single path
+    :param list package_configs: the list of optional package configurations submitted by the user
+    :raises FileAccessError: if a file is not present for the gomod package manager
+    :raises UnsupportedFeature: if dependency replacements are provided for
+        a non-single go module path
+    :raises GoModError: if failed to fetch gomod dependencies
+    """
+    Config.cachito_bundles_dir = request.output_dir
+
+    version_output = run_cmd(["go", "version"], {})
+    log.info(f"Go version: {version_output.strip()}")
+
+    config = get_worker_config()
```

Comment:
Should we have a global ("singleton"?) config object or just pass a config around as argument?

FWIW I think having a singleton is not a _terrible_ idea, as long as we don't couple everything to it